src/main/java/daehun/trip_java/User/service/CustomOAuth2UserService.java

```diff
@@ -0,0 +1,39 @@
+package daehun.trip_java.User.service;
+
+import daehun.trip_java.User.domain.User;
+import daehun.trip_java.User.repository.UserRepository;
+import daehun.trip_java.User.security.CustomOAuth2User;
+import java.util.Optional;
+import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+
+public class CustomOAuth2UserService extends DefaultOAuth2UserService {
+  private final UserRepository userRepository;
+
+  public CustomOAuth2UserService(UserRepository userRepository) {
+    this.userRepository = userRepository;
+  }
+
+  @Override
+  public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
+    OAuth2User oAuth2User = super.loadUser(userRequest);
+
+    String email = oAuth2User.getAttribute("email");
+    Optional<User> userOptional = userRepository.findByEmail(email);
+
+    User user;
+    if (userOptional.isPresent()) {
+      user = userOptional.get();
+    } else {
+      // 새 사용자 등록
+      user = new User();
+      user.setUsername(oAuth2User.getAttribute("name"));
+      user.setEmail(email);
+      userRepository.save(user);
+    }
```

Comment:
Optional type으로 변수를 받으면 가독성이 떨어지게 됩니다. 다음과 같이 하시면 좋을 것 같아요. 
```suggestion
    User user = userRepository.findByEmail(email).orElseGet(() -> {
        User newUser = new User();
        newUser.setUsername(oAuth2User.getAttribute("name"));
        newUser.setEmail(email);
        return userRepository.save(newUser);
    }
```
그리고 new User를 만드는 부분을 별도의 메소드로 빼면 좀더 깔끔해질 것 같네요.

Reply:
orElseGet 로 코드 간결화 하고, 별도 메서드로 분리해서 가독성 높이겠습니다.